// prettier-plugin-angular/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  default: () => index_default,
  options: () => options,
  parsers: () => parsers,
  printers: () => printers
});
module.exports = __toCommonJS(index_exports);

// src/parser.ts
var import_angular_html_parser = require("angular-html-parser");
function isElement(node) {
  return node.constructor?.name === "Element";
}
function isText(node) {
  return node.constructor?.name === "Text";
}
function isComment(node) {
  return node.constructor?.name === "Comment";
}
function isBlock(node) {
  return node.constructor?.name === "Block";
}
function normalizeNode(node, text) {
  if (isElement(node)) {
    const selfClose = node.endSourceSpan != null && node.startSourceSpan?.end?.offset === node.endSourceSpan?.end?.offset;
    return {
      type: "element",
      name: node.name,
      attrs: node.attrs.map(
        (a) => ({
          name: a.name,
          value: a.value ?? "",
          rawSource: text.slice(a.sourceSpan.start.offset, a.sourceSpan.end.offset),
          sourceSpan: a.sourceSpan,
          nameSpan: a.nameSpan ?? a.keySpan,
          valueSpan: a.valueSpan
        })
      ),
      children: (node.children ?? []).map((child) => normalizeNode(child, text)),
      sourceSpan: node.sourceSpan,
      startSourceSpan: node.startSourceSpan,
      endSourceSpan: selfClose ? null : node.endSourceSpan ?? null
    };
  }
  if (isText(node)) {
    return {
      type: "text",
      value: node.value,
      sourceSpan: node.sourceSpan
    };
  }
  if (isComment(node)) {
    return {
      type: "comment",
      value: node.value ?? null,
      sourceSpan: node.sourceSpan
    };
  }
  if (isBlock(node)) {
    return {
      type: "block",
      name: node.name,
      children: (node.children ?? []).map((child) => normalizeNode(child, text)),
      sourceSpan: node.sourceSpan,
      startSourceSpan: node.startSourceSpan ?? node.sourceSpan,
      endSourceSpan: node.endSourceSpan ?? null
    };
  }
  return {
    type: "expansion",
    sourceSpan: node.sourceSpan
  };
}
function parseAngularHtml(text) {
  const result = (0, import_angular_html_parser.parse)(text, { canSelfClose: true, tokenizeAngularBlocks: true });
  if (result.errors.length > 0) {
    console.warn("angular-html-parser errors:", result.errors);
    return { type: "root", nodes: [], originalText: text, hasErrors: true };
  }
  return {
    type: "root",
    nodes: result.rootNodes.map((n) => normalizeNode(n, text)),
    originalText: text
  };
}
var parsers = {
  "angular-attributes": {
    parse: parseAngularHtml,
    astFormat: "angular-attributes-ast",
    locStart: (node) => node.sourceSpan?.start?.offset ?? 0,
    locEnd: (node) => node.sourceSpan?.end?.offset ?? 0
  }
};

// src/attribute-sorter.ts
function getAttributeCategory(attr) {
  const name = attr.name;
  if (name.startsWith("#") || name.startsWith("let-")) return "ref";
  if (name.startsWith("*")) return "structural";
  if (name.startsWith("[(") && name.endsWith(")]")) return "twoWay";
  if (name.startsWith("[@")) return "animation";
  if (name.startsWith("[") && name.endsWith("]")) return "input";
  if (name.startsWith("(") && name.endsWith(")")) return "output";
  if (attr.value === "") return "boolean";
  return "static";
}
function sortAttributes(attrs, order) {
  const priority = (a) => {
    const cat = getAttributeCategory(a);
    const idx = order.indexOf(cat);
    return idx === -1 ? order.length : idx;
  };
  return [...attrs].sort((a, b) => priority(a) - priority(b));
}

// src/printer.ts
var VOID_ELEMENTS = /* @__PURE__ */ new Set([
  "area",
  "base",
  "br",
  "col",
  "embed",
  "hr",
  "img",
  "input",
  "link",
  "meta",
  "param",
  "source",
  "track",
  "wbr"
]);
function canSelfClose(tagName) {
  return VOID_ELEMENTS.has(tagName) || tagName.includes("-");
}
function serializeAttributeCollapsed(attr) {
  if (attr.value === "") return attr.name;
  const normalizedValue = attr.value.replace(/\s+/g, " ").trim();
  return `${attr.name}="${normalizedValue}"`;
}
function serializeAttribute(attr, attrIndent, tabWidth) {
  if (attr.value === "") return attr.name;
  if (attr.value.includes("\n")) {
    if (attrIndent !== void 0 && tabWidth !== void 0) {
      return serializeMultilineAttribute(attr, attrIndent, tabWidth);
    }
    return attr.rawSource;
  }
  const normalizedValue = attr.value.replace(/\s+/g, " ").trim();
  return `${attr.name}="${normalizedValue}"`;
}
function serializeMultilineAttribute(attr, attrIndent, tabWidth) {
  const lines = attr.value.split("\n");
  const rawContent = lines.slice(1, -1).filter((l) => l.trim().length > 0);
  const contentIndent = attrIndent + " ".repeat(tabWidth);
  let reindented;
  if (rawContent.length === 0) {
    reindented = [];
  } else {
    const minIndent = Math.min(...rawContent.map((l) => l.match(/^(\s*)/)?.[1].length ?? 0));
    reindented = rawContent.map((l) => contentIndent + l.slice(minIndent));
  }
  return [`${attr.name}="`, ...reindented, `${attrIndent}"`].join("\n");
}
function measureSingleLine(tagName, attrs, indent, selfClose) {
  const closing = selfClose ? " />" : ">";
  let len = indent.length + 1 + tagName.length;
  for (const a of attrs) {
    len += 1 + serializeAttributeCollapsed(a).length;
  }
  len += closing.length;
  return len;
}
function formatDocument(root, opts) {
  if (root.hasErrors) return root.originalText;
  return formatNodes(root.nodes, root.originalText, "", opts);
}
function formatNodes(nodes, originalText, indent, opts) {
  return nodes.map((node) => formatNode(node, originalText, indent, opts)).join("");
}
function formatNode(node, originalText, indent, opts) {
  if (node.type === "element") return formatElement(node, originalText, indent, opts);
  if (node.type === "block") return formatBlock(node, originalText, indent, opts);
  return originalText.slice(node.sourceSpan.start.offset, node.sourceSpan.end.offset);
}
function formatBlock(node, originalText, indent, opts) {
  const header = originalText.slice(node.startSourceSpan.start.offset, node.startSourceSpan.end.offset);
  const footer = node.endSourceSpan ? originalText.slice(node.endSourceSpan.start.offset, node.endSourceSpan.end.offset) : "}";
  const childIndent = indent + " ".repeat(opts.tabWidth);
  const children = formatChildren(node.children, originalText, childIndent, opts);
  return children ? `${indent}${header}
${children}
${indent}${footer}` : `${indent}${header}
${indent}${footer}`;
}
function formatChildren(nodes, originalText, baseIndent, opts) {
  const results = [];
  for (const child of nodes) {
    if (child.type === "text") {
      const raw = originalText.slice(child.sourceSpan.start.offset, child.sourceSpan.end.offset);
      for (const raw_line of raw.split("\n")) {
        const line = raw_line.trim();
        if (line) results.push(baseIndent + line);
      }
    } else {
      results.push(formatNode(child, originalText, baseIndent, opts));
    }
  }
  return results.filter((s) => s !== "").join("\n");
}
function formatElement(node, originalText, indent, opts) {
  const attrs = opts.angularAttributeSort ? sortAttributes(node.attrs, opts.angularAttributeOrder) : node.attrs;
  const allChildrenWhitespace = node.children.length > 0 && node.children.every(
    (child) => child.type === "text" && originalText.slice(child.sourceSpan.start.offset, child.sourceSpan.end.offset).trim() === ""
  );
  const hasChildren = node.children.length > 0 && !allChildrenWhitespace;
  const selfClose = !hasChildren && canSelfClose(node.name);
  const singleLineLen = measureSingleLine(node.name, attrs, indent, selfClose);
  const tagPrefix = `<${node.name} `;
  const tagPrefixLen = indent.length + tagPrefix.length;
  const firstAttrLen = attrs.length > 0 ? serializeAttributeCollapsed(attrs[0]).length : 0;
  const closingLen = selfClose ? 3 : 1;
  const caseBFirstLineLen = tagPrefixLen + firstAttrLen + (attrs.length === 1 ? closingLen : 0);
  const hasMultilineAttr = attrs.some((a) => a.value.includes("\n"));
  let openTag;
  if (!hasMultilineAttr && singleLineLen <= opts.printWidth) {
    openTag = formatOpenTagInline(node.name, attrs, selfClose);
  } else if (tagPrefixLen < opts.printWidth && caseBFirstLineLen <= opts.printWidth) {
    const firstAttrCol = tagPrefixLen;
    openTag = formatOpenTagAligned(node.name, attrs, firstAttrCol, selfClose, opts.tabWidth);
  } else {
    openTag = formatOpenTagFallback(node.name, attrs, indent, opts.tabWidth, selfClose);
  }
  if (!hasChildren) {
    if (selfClose) return indent + openTag;
    return `${indent}${openTag}</${node.name}>`;
  }
  const singleTextChild = node.children.length === 1 && node.children[0].type === "text" ? originalText.slice(node.children[0].sourceSpan.start.offset, node.children[0].sourceSpan.end.offset).trim() : null;
  if (singleTextChild !== null) {
    const totalLen = singleLineLen + singleTextChild.length + node.name.length + 3;
    if (totalLen <= opts.printWidth) {
      return `${indent}${openTag}${singleTextChild}</${node.name}>`;
    }
  }
  const childIndent = indent + " ".repeat(opts.tabWidth);
  const children = formatChildren(node.children, originalText, childIndent, opts);
  return `${indent}${openTag}
${children}
${indent}</${node.name}>`;
}
function formatOpenTagInline(name, attrs, selfClose) {
  const attrsStr = attrs.map((a) => serializeAttribute(a)).join(" ");
  const closing = selfClose ? " />" : ">";
  return attrsStr ? `<${name} ${attrsStr}${closing}` : `<${name}${closing}`;
}
function formatOpenTagAligned(name, attrs, firstAttrCol, selfClose, tabWidth) {
  if (attrs.length === 0) return selfClose ? `<${name} />` : `<${name}>`;
  const padding = " ".repeat(firstAttrCol);
  const closing = selfClose ? " />" : ">";
  const [first, ...rest] = attrs;
  if (rest.length === 0) {
    return `<${name} ${serializeAttribute(first, padding, tabWidth)}${closing}`;
  }
  const restLines = rest.map((a, i) => {
    const isLast = i === rest.length - 1;
    return padding + serializeAttribute(a, padding, tabWidth) + (isLast ? closing : "");
  });
  return [`<${name} ${serializeAttribute(first, padding, tabWidth)}`, ...restLines].join("\n");
}
function formatOpenTagFallback(name, attrs, indent, tabWidth, selfClose) {
  if (attrs.length === 0) return selfClose ? `<${name} />` : `<${name}>`;
  const attrIndent = indent + " ".repeat(tabWidth);
  const closing = selfClose ? " />" : ">";
  const attrLines = attrs.map((a, i) => {
    const isLast = i === attrs.length - 1;
    return attrIndent + serializeAttribute(a, attrIndent, tabWidth) + (isLast ? closing : "");
  });
  return [`<${name}`, ...attrLines].join("\n");
}
var printers = {
  "angular-attributes-ast": {
    print(path, options2, _print) {
      const root = path.getValue();
      if (root.type !== "root") return "";
      return formatDocument(root, options2);
    }
  }
};

// src/options.ts
var options = {
  angularAttributeSort: {
    type: "boolean",
    category: "Angular",
    default: true,
    description: "Sort Angular HTML attributes by category (ref, structural, twoWay, input, output, animation, static, boolean)"
  },
  angularAttributeOrder: {
    type: "string",
    category: "Angular",
    array: true,
    default: [
      {
        value: [
          "ref",
          "structural",
          "twoWay",
          "input",
          "output",
          "animation",
          "static",
          "boolean"
        ]
      }
    ],
    description: "Order of attribute categories for sorting"
  }
};

// src/index.ts
var index_default = { parsers, printers, options };
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  options,
  parsers,
  printers
});
